analysis/topology/topological_space.lean

```diff
@@ -634,37 +634,377 @@ e ▸ this
+    { rcases ih S (finset.subset.refl S) hs with ⟨y, hy, V, hv1, hv2⟩,
+      by_cases hxV : x ∈ V,
+      { cases t0_space.t0 x y (λ hxy, hxS $ by rwa hxy) with U hu,
+        rcases hu with ⟨hu1, ⟨hu2, hu3⟩ | ⟨hu2, hu3⟩⟩,
+        { existsi [x, finset.mem_insert_self x S, U ∩ V, is_open_inter hu1 hv1],
+          apply set.ext,
+          intro z,
+          split,
+          { intro hzx,
+            rw set.mem_singleton_iff at hzx,
+            rw hzx,
+            exact ⟨finset.mem_insert_self x S, ⟨hu2, hxV⟩⟩ },
+          { intro hz,
+            rw set.mem_singleton_iff,
+            rcases hz with ⟨hz1, hz2, hz3⟩,
+            cases finset.mem_insert.1 hz1 with hz4 hz4,
+            { exact hz4 },
+            { have h1 : z ∈ {y : α | y ∈ S} ∩ V,
+              { exact ⟨hz4, hz3⟩ },
+              rw ← hv2 at h1,
+              rw set.mem_singleton_iff at h1,
+              rw h1 at hz2,
+              exact (hu3 hz2).elim } } },
+        { existsi [y, finset.mem_insert_of_mem hy, U ∩ V, is_open_inter hu1 hv1],
+          apply set.ext,
+          intro z,
+          split,
+          { intro hz,
+            rw set.mem_singleton_iff at hz,
+            rw hz,
+            refine ⟨finset.mem_insert_of_mem hy, hu2, _⟩,
+            have h1 : y ∈ {y} := set.mem_singleton y,
+            rw hv2 at h1,
+            exact h1.2 },
+          { intro hz,
+            rw set.mem_singleton_iff,
+            cases hz with hz1 hz2,
+            cases finset.mem_insert.1 hz1 with hz3 hz3,
+            { rw hz3 at hz2,
+              exact (hu3 hz2.1).elim },
+            { have h1 : z ∈ {y : α | y ∈ S} ∩ V := ⟨hz3, hz2.2⟩,
+              rw ← hv2 at h1,
+              rw set.mem_singleton_iff at h1,
+              exact h1 } } } },
+      { existsi [y, finset.mem_insert_of_mem hy, V, hv1],
+        apply set.ext,
+        intro z,
+        split,
+        { intro hz,
+          rw set.mem_singleton_iff at hz,
+          rw hz,
+          split,
+          { exact finset.mem_insert_of_mem hy },
+          { have h1 : y ∈ {y} := set.mem_singleton y,
+            rw hv2 at h1,
+            exact h1.2 } },
+        { intro hz,
+          rw hv2,
+          cases hz with hz1 hz2,
+          cases finset.mem_insert.1 hz1 with hz3 hz3,
+          { rw hz3 at hz2,
+            exact (hxV hz2).elim },
+          { exact ⟨hz3, hz2⟩ } } } } }
+end,
+begin
+  apply nonempty.elim ha, intro x,
+  specialize H finset.univ (finset.ne_empty_of_mem $ finset.mem_univ x),
+  rcases H with ⟨y, hyf, U, hu1, hu2⟩,
+  existsi y,
+  have h1 : {y : α | y ∈ finset.univ} = (univ : set α),
+  { exact set.eq_univ_of_forall (λ x : α,
+      by rw mem_set_of_eq; exact finset.mem_univ x) },
+  rw h1 at hu2,
+  rw set.univ_inter at hu2,
+  rw hu2,
+  exact hu1
+end
+
 /-- A T₁ space, also known as a Fréchet space, is a topological space
   where for every pair `x ≠ y`, there is an open set containing `x` and not `y`.
   Equivalently, every singleton set is closed. -/
```

Comment:
Now would be a good time to fix this docstring (actually referring to the Lean-implemented definition first) and then state and prove in Lean the equivalence.